### PR TITLE
Hyperparameter tuning and testing

### PR DESCRIPTION
**Pull request by mkosturek**
W sumie to to jest najświeższy stan repo - tu były robione badania i ostatnie tuningi do oddania projektu, tyle że po fakcie zapomniałem zrobić PR.

Przejrzyjcie kod mimo wszystko, żeby nie okazało się, że ten PR usunie coś z czego korzystacie itp itd...

EDIT: zignorujcie proszę literówkę w wiadomości do commita d4d852c :P

### CODE REVIEW

File: tweet_recommendations/datasets_access.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    hashtags = [{'hashtag': node, **G.nodes[node]}
+                for node in G.nodes
+                if G.nodes[node]['node_type'] == 'hashtag']
+    hashtags = pd.DataFrame(hashtags)
+    return hashtags
+
+
+def add_hashtag_occurences_counts(hashtags_df, reference_tweets_df):
+    all_hashtags = []
+    _ = reference_tweets_df.hashtags.apply(lambda h: [all_hashtags.append(x) for x in set(h)])
```
Comment (jmichalowski12) [inline tweet_recommendations/datasets_access.py]: Czy musimy tutaj wrzucać "_" ? Samo wywołanie chyba wystarczy

File: tweet_recommendations/datasets_access.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    cnts = Counter(all_hashtags)
```
Comment (jmichalowski12) [inline tweet_recommendations/datasets_access.py]: Może pierdoła, ale 2 sekundy zastanawiałem się czym jest cnts :D Jak by ci się chciało to można nazwać zmienną hashtag_counts?

File: tweet_recommendations/datasets_access.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+def get_datasets():
+    full_dataset = get_full_dataset()
+    train_df, train_pop, train_unpop = get_train_dataset(full_dataset)
+    val_df, val_pop, val_unpop = get_val_dataset(full_dataset)
+    test_df, test_pop, test_unpop = get_test_dataset(full_dataset)
+    return train_df, val_df, test_df, test_pop, test_unpop
+
+
+def get_hashtags_df(graph_path='/data/graphs/train_graph.p'):
```
Comment (jmichalowski12) [inline tweet_recommendations/datasets_access.py]: Byłbym ultra szczęśliwy, gdyby było opisane co ten dataframe zawiera po transformacji z grafu na dataframe :D - nie koniecznie super dokładnie ale tak, żeby nie musieć szukać gdzieś co tam w środku wszystko siedziało

File: tweet_recommendations/recommendation.py [hunk 4]
```diff
@@ -11,32 +11,46 @@
...
     ang_dist = np.arccos(similarity) / np.pi
     ang_sim = 1 - ang_dist
     return ang_sim
 
 
 def normalise(array: np.ndarray):
-    return (array - array.min()) / (array.max() - array.min())
+    if array.max() != array.min():
+        return (array - array.min()) / (array.max() - array.min())
+    else:
```
Comment (jmichalowski12) [inline tweet_recommendations/recommendation.py]: Ten else chyba nie jest potrzebny?

Btw ta normalizacja powinna być w przedziale [0, 1]? Jeśli jednak [-1, 1] to niżej mamy sampla: 

> array = array - array.mean()
>array = array / array.max()

**Review by jmichalowski12 — COMMENTED** [review]
Ogólnie jest bardzo gites 🥇 -> prawo i porządek 

Dwie w sumie uwagi generalne: 
1) ponawiam propozycję formatowania [blackiem](https://github.com/ambv/black) kodu - uprości nam to troszkę pracę (i nie będzie dywagacji, gdzie złamać linię itp)
2) Dobrze mi się czyta ten kod bo bardzo fajnie opisujesz zmienne, nie mniej jednak w dłuższych metodkach typ argumentów/return type w sygnaturze metody byłby sztosik, zwłaszcza, kiedy typ nie jest zawarty w nazwie zmiennej :)

**Review by jmichalowski12 — APPROVED** [review]

File: tweet_recommendations/recommendation.py [hunk 5]
```diff
@@ -11,32 +11,46 @@
...
-    hashtag_embeddings = np.vstack(hashtags_df[embedding_name])
-    similarities = embedding_similarity(embedding.reshape(1, -1), hashtag_embeddings).reshape(-1)
-
+    similarities = similarities.reshape(-1)
     sim_pop_mix = prepare_similarity_and_popularity_mix(similarities,
                                                         hashtags_df[popularity_measure].values,
                                                         popularity_to_similarity_ratio, config)
 
-    top_k = sim_pop_mix.argsort()[-K:][::-1]
+    top_k = np.argpartition(-sim_pop_mix, np.arange(K))[:K]
```
Comment (kacper1095) [inline tweet_recommendations/recommendation.py:60]: Wydaje mi się, że czytelność może zwiększyć po prostu `np.argsort(-sim_pop_mix[:K])`

File: tweet_recommendations/datasets_access.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    cnts = pd.DataFrame(list(cnts.items()), columns=['hashtag', 'count'])
+
+    return hashtags_df.merge(cnts, on='hashtag')
+
+
+def filter_hashtags_by_occurences_limit(hashtags_df, limit):
+    return hashtags_df[hashtags_df['count'] > limit]
+
+
+def compute_expected_hashtags_with_respect_to_hashtags_df(tweets_df, hashtags_df):
```
Comment (kacper1095) [inline tweet_recommendations/datasets_access.py]: Nazwa metody trochę nieadekwantna do tego, co się dzieje, prędzej coś w stylu `filter_tweets_without_expected_hashtags(tweets_df, expected_hashtags_df)`

File: tweet_recommendations/grid_search.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+def create_progressbar():
+    return ProgressBar()
+
+
+def grid_search(datasets_pool, hashtags_df, hyperparameters_df, metrics_to_calc=METRICS, backup_dir=None, progressbar=suppress()):
+    if backup_dir is not None:
+        os.makedirs(backup_dir, exist_ok=True)
+
+    bag = db.from_sequence([hyperparameters_df.iloc[i].to_dict()
```
Comment (kacper1095) [inline tweet_recommendations/grid_search.py]: Wyczytałem https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_dict.html, że można po prostu `hyperparameters_df.to_dict('records')`  <- to zwraca listę słowników

File: tweet_recommendations/recommendation.py [hunk 8]
```diff
@@ -11,32 +11,46 @@
 
 
 def embedding_similarity(x: np.ndarray, y: np.ndarray):
-    similarity = cosine_similarity(x, y)
+    similarity = np.clip(cosine_similarity(x, y), -1, 1)
```
Comment (kacper1095) [inline tweet_recommendations/recommendation.py:21]: Czy cosine similarity jest dobrze liczony skoro tutaj upewniasz się, że jest -1 do 1?

File: tweet_recommendations/recommendation.py [hunk 9]
```diff
@@ -11,32 +11,46 @@
...
+        hashtag_embeddings = config['override_hashtag_embeddings'][embedding_name]
+    else:
+        hashtag_embeddings = np.asarray(hashtags_df[embedding_name].values.tolist())
+    similarities = embedding_similarity(embedding.reshape(1, -1), hashtag_embeddings).reshape(-1)
+
+    return recommend_with_computed_similarities(similarities, hashtags_df, config)
+
+
+def recommend_with_computed_similarities(similarities: np.ndarray, hashtags_df: pd.DataFrame, config: dict):
+    assert all(key in config for key in CONFIG_KEYS if key != 'w2v_function')
```
Comment (kacper1095) [inline tweet_recommendations/recommendation.py]: Jeśli jest taka asercja, to przydałby się komentarz albo wiadomość, którą wypluwa niespełnia asercja, w stylu `assert <condition>, "<message>"`. Rozumiem jego przeznaczenie do momentu ifa `key != 'w2v_function'`. To samo metoda wyżej.

File: tweet_recommendations/recommendation.py [hunk 10]
```diff
@@ -11,32 +11,46 @@
...
-    similarity = cosine_similarity(x, y)
+    similarity = np.clip(cosine_similarity(x, y), -1, 1)
     ang_dist = np.arccos(similarity) / np.pi
     ang_sim = 1 - ang_dist
     return ang_sim
 
 
 def normalise(array: np.ndarray):
-    return (array - array.min()) / (array.max() - array.min())
+    if array.max() != array.min():
```
Comment (kacper1095) [inline tweet_recommendations/recommendation.py]: Jeśli są to floaty, to warunek będzie prawdziwy co najwyżej z przypadku (na wbudowanym typie floata chyba nie ma takich przypadków). Z drugiej strony różnica może dać 0 przez co poniższe się wysypie. To jest odczucie, nie musi się wcale tak zdarzyć, ale proponowałbym użycie `np.isclose(array.max(), array.min(), rtol=1e-8)`.

File: tweet_recommendations/grid_search.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                       axis=1)
+    result = pd.DataFrame(list(metrics)).mean()
+
+    del df, hashtags, precomputed_similarities
+
+    if backup_dir is not None:
+        bak = {col: result[col] for col in result.index}
+        fname = '_'.join([str(x) for x in hyperparameters.values()]) + '.pkl'
+        with open(backup_dir + '/' + fname, 'wb') as f:
+            pickle.dump({**hyperparameters, **bak}, f)
```
Comment (kacper1095) [inline tweet_recommendations/grid_search.py:85]: image

File: tweet_recommendations/graphs/graph_builder.py [hunk 12]
```diff
@@ -100,11 +100,11 @@ def calculate_hashtag_popularity_mean_retweets_heuristic(G: nx.Graph, progress_b
...
         if G.nodes[node]["node_type"] == "hashtag":
             tweets = G.neighbors(node)
-            retweets_counts = np.asarray([G.node[tweet]['retweets'] for tweet in tweets])
-            G.node[node]['mean_retweets'] = retweets_counts.mean(axis=0)
+            retweets_counts = np.asarray([G.nodes[tweet]['retweets'] for tweet in tweets])
+            G.nodes[node]['mean_retweets'] = retweets_counts.mean(axis=0)
 
     for node in G.nodes:
-        assert 'mean_retweets' in G.node[node] if G.node[node]['node_type'] == 'hashtag' else True
+        assert 'mean_retweets' in G.nodes[node] if G.nodes[node]['node_type'] == 'hashtag' else True
```
Comment (kacper1095) [inline tweet_recommendations/graphs/graph_builder.py]: Tak samo jak wcześniej (albo później), preferowałbym asercję z wiadomością albo komentarzem. Jeśli się program wywali na nim, to będę tyle samo wiedział, co w tym momencie

File: tweet_recommendations/datasets_access.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
+import os
+import pandas as pd
+import pickle
+from collections import Counter
+
+
+PATH = os.path.dirname(os.path.abspath(__file__)) + '/../'
```
Comment (kacper1095) [inline tweet_recommendations/datasets_access.py]: Jak chcemy być pro cross platform, to możemy wszędzie używać `os.path.join`, żeby łączenie ścieżek prawidłowe. O ile takie łączenie działa na windowsie, tak np. windowsowe `\\path\\to\\something` nie działa na linuxie. Polecam też `from pathlib import Path`, ale to na przyszłość :)

File: tweet_recommendations/grid_search.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+    df['recommendations'] = df['similarities'].apply(
+        lambda sim: recommend_with_computed_similarities(sim, hashtags, config))
+
+    metrics = df.apply(lambda row: get_metrics_for_recommendation(row['expected_hashtags'], row['recommendations'],
+                                                                  metrics_to_calculate, config['K']),
+                       axis=1)
+    result = pd.DataFrame(list(metrics)).mean()
+
+    del df, hashtags, precomputed_similarities
```
Comment (kacper1095) [inline tweet_recommendations/grid_search.py]: Prawdopodobnie to nie wpływa w żaden sposób na program, zwłaszcza, że za chwilę następuje powrót z metody.

**Review by kacper1095 — APPROVED** [review]

**Comment by kacper1095** [comment]
Polecam "Ctrl + Alt + Shift + L" w pycharmie i wpisywanie wszędzie typów, jeśli już w niektórych miejscach się pojawia ;)

File: tweet_recommendations/recommendation.py (hunk 5 above)
Comment (mkosturek) [reply tweet_recommendations/recommendation.py:60]: Czytelność tak, ale szybkość zdecydowanie nie. Do robienia badań/grid searcha to miało znaczenie. Argpartition wybierze top K hashtagów, argsort będzie musiał przesortować całą listę tysięcy hashtagów.

File: tweet_recommendations/recommendation.py (hunk 4 above)
Comment (mkosturek) [reply tweet_recommendations/recommendation.py]: Nie wiem chyba dlaczego normalizacja miałaby być do [-1, 1]...
Else wyrzucony

File: tweet_recommendations/datasets_access.py (hunk 2 above)
Comment (mkosturek) [reply tweet_recommendations/datasets_access.py]: Załatwione

File: tweet_recommendations/datasets_access.py (hunk 1 above)
Comment (mkosturek) [reply tweet_recommendations/datasets_access.py]: Zrobione.

File: tweet_recommendations/grid_search.py (hunk 7 above)
Comment (mkosturek) [reply tweet_recommendations/grid_search.py]: OK, dobrze wiedzieć. 

Zamykam wątek. 
mk
